Quantum-Control-Applications/Superconducting/Two-Flux-Tunable-Coupled-Transmons/Use Case 4 - Single- and Two-Qubit State and Process Tomography/single-qubit-process-tomography.py

```diff
@@ -0,0 +1,245 @@
+from qm import QuantumMachinesManager
+from qm import SimulationConfig
+from configuration import *
+from qualang_tools.results import progress_counter, fetching_tool
+import numpy as np
+from scipy.linalg import solve
+
+from helper_functions import (
+    P_Pauli1,
+    plot_process_tomography1,
+    map_from_bloch_state_to_pauli_basis1,
+    func_E1,
+)
+
+
+###################
+# The QUA program #
+###################
+
+# qubit under test, assuming there are multiple qubits
+# with XY line elements "q<qubit>_xy", flux lines
+# "q<qubit>_z", and readout resonator elements "rr<qubit>"
+qubit = 1
+
+if qubit == 1:
+    threshold = ge_threshold_q1
+elif qubit == 2:
+    threshold = ge_threshold_q2
+else:
+    raise ValueError(f"Incorrect qubit number chosen")
+
+
+n_avg = 10_000
+
+
+# subroutine to prepare desired qubit gate/process
+def analysed_process(qubit):
+    # write whatever QUA code you need, here, in order to perform the
+    # desired process which we want to subject to tomography.
+    # For example, to analyse the Y gate:
+    play("y180", f"q{qubit}_xy")
+
+
+ideal_gate = func_E1(2)
+
+
+# subroutine to switch between preparing the tomographically-complete
+# set of input states
+def prepare_state(i):
+    with switch_(i):
+        with case_(0):
+            wait(pi_len // 4, f"q{qubit}_xy")
+        with case_(1):
+            play("x180", f"q{qubit}_xy")
+        with case_(2):
+            play("y90", f"q{qubit}_xy")
+        with case_(3):
+            play("-y90", f"q{qubit}_xy")
+        with case_(4):
+            play("-x90", f"q{qubit}_xy")
+        with case_(5):
+            play("x90", f"q{qubit}_xy")
+
+
+# subroutine to switch between the tomographically-complete
+# set of bases in which to measure
+def measurement_basis_change(j):
+    with switch_(j):
+        with case_(0):
+            wait(pi_len // 4, f"q{qubit}_xy")
+        with case_(1):
+            play("x180", f"q{qubit}_xy")
+        with case_(2):
+            play("y90", f"q{qubit}_xy")
+        with case_(3):
+            play("-y90", f"q{qubit}_xy")
+        with case_(4):
+            play("-x90", f"q{qubit}_xy")
+        with case_(5):
+            play("x90", f"q{qubit}_xy")
+
+
+with program() as single_qubit_process_tomography:
+
+    n = declare(int)
+    n_st = declare_stream()
+
+    I = declare(fixed)
+    Q = declare(fixed)
+
+    state = declare(bool)  # QUA variable for the measured qubit state
+    state_st = declare_stream()  # Stream for the qubit state
+
+    c = declare(int)  # QUA variable for switching between state preparation/creations
+    m = declare(int)  # QUA variable for switching between basis projections/measurements
+
+    with for_(n, 0, n < n_avg, n + 1):  # QUA for_ loop for averaging
+        with for_(c, 0, c <= 5, c + 1):  # QUA for_ loop for switching between state preparations
+            with for_(m, 0, m <= 5, m + 1):  # QUA for_ loop for switching between basis projections/measurements
+
```

Comment:
@bornman-nick there is a duplicate of `reset_phase()` here. Also, is there a need to `reset_phase()` that will bring each shot to shot to have identical phases. That means that from the beginnin of the program the qubit's bloch sphere is rotating freely but we are reseting the frame of the OPX oscillator every time. As a result, the X-gate (for example) is not the same as it was in the previous shot. However, if we don't reset-phase then the X-gate the program applies in the first shot, it is the same as the one in the last shot of the experiment because both the OPX and the qubit are rotating coherently.

Having said that what is the experimental requirement?

Reply:
@KevinAVR you are right, I added the `reset_frame` and `reset_phase` after testing the programme itself (I mindlessly copied and pasted it from another programme). We can take them both out.